File: Java/67. Add Binary.java
class Solution {
    public String addBinary(String a, String b) {
        int carry = 0;
        int i = 0;
        ArrayList<Integer> ans = new ArrayList<Integer>();
        StringBuilder sa = new StringBuilder(a);
        StringBuilder sb = new StringBuilder(b);
        sa.reverse();
        sb.reverse();
        
        while(i<a.length() || i<b.length()){
            int sum;
            
            if(i >= a.length()){
                sum = sb.charAt(i)-'0' + carry;
            }
            else if(i >= b.length()){
                sum = sa.charAt(i)-'0' + carry;
            }
            else{
                sum = sa.charAt(i)-'0' + sb.charAt(i)-'0' + carry;
            }
            
            carry = sum/2;
            int res = sum%2;
            
            ans.add(res);
            i++;
        }
        
        if(carry > 0){
            ans.add(carry);
        }
        
        StringBuilder res = new StringBuilder();
        for(i=0; i<ans.size(); i++){
            res.append(ans.get(i));
        }
        
        return res.reverse().toString();
    }
}

File: Python3/67. Add Binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        carry = 0
        i = 0
        ans = []
        
        a = a[::-1]
        b = b[::-1]
        
        # the result is reversed
        while i<len(a) or i<len(b):
            
            if i>=len(a):
                sum = int(b[i]) + carry
                
            elif i>=len(b):
                sum = int(a[i]) + carry
                
            else:
                sum = int(a[i]) + int(b[i]) + carry
            
            carry = floor(sum/2)
            res = sum%2
            
            ans.append(res)
            i += 1
            
        if carry:
            ans.append(carry)
            
        ans.reverse()
        
        return ''.join(str(digit) for digit in ans)